pkg/dynamicconfig/table.go

```diff
@@ -0,0 +1,128 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package dynamicconfig
+
+import (
+	"sort"
+	"strconv"
+
+	"github.com/cilium/hive/job"
+	"github.com/cilium/statedb"
+	"github.com/cilium/statedb/index"
+
+	"github.com/cilium/cilium/pkg/k8s"
+)
+
+const TableName = "cilium-configs"
+
+var (
+	keyIndex = statedb.Index[DynamicConfig, Key]{
+		Name: "config-name-source-key",
+		FromObject: func(t DynamicConfig) index.KeySet {
+			return index.NewKeySet(index.Stringer(t.Key))
+		},
+		FromKey: index.Stringer[Key],
+		Unique:  true,
+	}
+
+	ByKeyIndex = keyIndex.Query
+
+	keyNameIndex = statedb.Index[DynamicConfig, string]{
+		Name: "config-entry-key",
+		FromObject: func(t DynamicConfig) index.KeySet {
+			return index.NewKeySet(index.String(t.Key.Name))
+		},
+		FromKey: index.String,
+		Unique:  false,
+	}
+	ByKeyNameIndex = keyNameIndex.Query
+)
+
+type Key struct {
+	Name   string
+	Source string
+}
+
+func (k Key) String() string {
+	return k.Name + "/" + k.Source
+}
+
+type DynamicConfig struct {
+	Key   Key
+	Value string
+}
+
+func (d DynamicConfig) TableHeader() []string {
+	return []string{
+		"Key",
+		"Source",
+		"Priority",
+		"Value",
+	}
+}
+
+func (d DynamicConfig) TableRow() []string {
+	return []string{
+		d.Key.Name,
+		d.Key.Source,
+		strconv.Itoa(priorities[d.Key.Source]),
+		d.Value,
+	}
+}
+
+func NewConfigTable(db *statedb.DB) (statedb.RWTable[DynamicConfig], error) {
+	tbl, err := statedb.NewTable(
+		TableName,
+		keyIndex,
+		keyNameIndex,
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	return tbl, db.RegisterTable(tbl)
+}
+
+func RegisterConfigMapReflector(jobGroup job.Group, db *statedb.DB, rcs []k8s.ReflectorConfig[DynamicConfig], c config) error {
+	if !c.EnableDynamicConfig {
+		return nil
+	}
+	for _, rc := range rcs {
+		if err := k8s.RegisterReflector[DynamicConfig](jobGroup, db, rc); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func GetKey(txn statedb.ReadTxn, table statedb.Table[DynamicConfig], key string) (DynamicConfig, bool) {
+	entries := statedb.Collect(table.List(txn, ByKeyNameIndex(key)))
+
+	if len(entries) == 0 {
+		return DynamicConfig{}, false
+	}
+
+	sortByPriority(entries)
+
+	return entries[0], true
+}
+
+func WatchKey(txn statedb.ReadTxn, table statedb.Table[DynamicConfig], key string) (DynamicConfig, <-chan struct{}) {
```

Comment:
Hm slightly inconsistent that `GetKey` returns `(DynamicConfig, bool)` and here we don't have the `bool`. Begs the question whether `GetKey` needs it either. I would make these consistent to each other.